### ISSUE TITLE
Recursor must allow hostnames in forward-zone to be Docker / Docker Swarm compatible

### ISSUE THREAD

**Issue by jnehlmeier** [PowerDNS/pdns#11393]
- Program: Authoritative, Recursor
 - Issue type: Feature request

### Description

When having a docker swarm cluster with several hosts you use an overlay network to let containers communicate across hosts. In combination with docker swarm ingress port publishing you can easily choose multiple docker swarm cluster hosts to act as nameserver and requests will always end up in the PowerDNS Recursor container, regardless of the hosts it is executed on.

So for example you can have `/etc/resolv.conf` with 
```
nameserver 10.1.1.1:53 # docker swarm host 1 of 5
nameserver 10.1.1.2:53 # docker swarm host 2 of 5
```
and have a recursor running once (executed on any of the 5 hosts).

When also having private DNS zones and deploying an authoritative server to docker swarm, the recursor needs an entry in its `forward-zone` configuration to point to that authoritative server for the private zones. In that configuration option only IP addresses are allowed.

Given that you do not know where the authoritative server will be executed by docker you need to use the docker service name of the authoritative server instead of an IP. That service name will point to a virtual IP provided by docker and let docker route the requests to the appropriate host/container.

So in short: Running recursor and authoritative inside docker swarm does effectively not work since you can not forward to the authoritative server as you do not know the IP. The only solution is to allow hostnames in addition to IPs in `forward-zone` which the docker embedded DNS server will resolve to the correct target IP.

**Comment by jnehlmeier** [comment]
An additional solution would be to provide a docker image which runs recursor and authoritative server in the same container so that both can talk to each other via localhost.

**Comment by Habbie** [comment]
Using DNS to run your DNS is a recipe for confusion and hard to debug problems. We recommend writing some tooling that generates/updates your forward-zones.

**Comment by Habbie** [comment]
> Using DNS to run your DNS is a recipe for confusion and hard to debug problems. We recommend writing some tooling that generates/updates your forward-zones.

`dnsdist-resolver.*` in https://github.com/PowerDNS/pdns/tree/master/dockerdata might provide some inspiration - although the Recursor does not provide the same convenient Lua hooks.

**Comment by omoerbeek** [comment]
One possible approach is to set forward zones using the REST API.

**Comment by jnehlmeier** [comment]
> 

Of course I could create a shell script as entrypoint for the container and before starting the recursor do a DNS lookup (through docker embedded DNS server) asking for the virtual IP of the authoritative server service. Then I could update the config and start the recursor. 

However there is no guarantee that the IP won't be changed by docker while the docker daemon is running. Maybe it can be assumed that it won't change until you restart the docker daemon but it feels brittle.

**Comment by omoerbeek** [comment]
I can see the use for this, but it's quite tricky to get right.

**Comment by jnehlmeier** [comment]
For now I am using a different DNS server which has a builtin web UI for configuration/zone administration and can act as authoritative server and recursor. So basically an all-in-one solution. Works pretty well so far.

But I'll keep an eye on pdns and this issue.

**Comment by frenzymind** [comment]
> One possible approach is to set forward zones using the REST API.

May I ask you to give example of code. When I explore docs of API I found this is not possible. It seems I don't understand docs correctly.

**Comment by omoerbeek** [comment]
https://docs.powerdns.com/recursor/http-api/endpoint-zones.html#post--api-v1-servers--server_id-zones
With a POST body as specified in 
https://docs.powerdns.com/recursor/http-api/zone.html#zone
I'll add an example here and to the docs.

**Comment by omoerbeek** [comment]
To create a forward (please read the notes about `forward-zone`, `include-dir` and `api-dir` in https://docs.powerdns.com/recursor/http-api/zone.html#zone
```
curl --no-progress-meter -H 'X-API-Key: secret' -H 'Content-type: application/json' -X POST --data-binary @- http://localhost:8082/api/v1/servers/localhost/zones << EOF | jq
{
  "name": "example.com.",
  "type": "Zone",
  "kind": "Forwarded",
  "servers": ["192.168.178.1", "192.168.178.2:5353"],
  "recursion_desired" : false
}
EOF
```
Output:
```
{
  "id": "example.com.",
  "kind": "Forwarded",
  "name": "example.com.",
  "records": [],
  "recursion_desired": false,
  "servers": [
    "192.168.178.1:53",
    "192.168.178.2:5353"
  ],
  "url": "/api/v1/servers/localhost/zones/example.com."
}
```
To delete the forward created above:
```
curl --no-progress-meter -H 'X-API-Key: secret' -H 'Content-type: application/json' -X DELETE http://localhost:8082/api/v1/servers/localhost/zones/example.com. << EOF
{
  "name": "example.com.",
  "type": "Zone",
  "kind": "Forwarded",
  "servers": ["192.168.178.1", "192.168.178.2:5353"],
  "recursion_desired" : false
}
EOF
```
